pkg/ottl/ottlfuncs/func_is_match.go

```diff
@@ -31,19 +31,35 @@ func createIsMatchFunction[K any](_ ottl.FunctionContext, oArgs ottl.Arguments)
 	return isMatch(args.Target, args.Pattern)
 }
 
-func isMatch[K any](target ottl.StringLikeGetter[K], pattern string) (ottl.ExprFunc[K], error) {
-	compiledPattern, err := regexp.Compile(pattern)
-	if err != nil {
-		return nil, fmt.Errorf("the pattern supplied to IsMatch is not a valid regexp pattern: %w", err)
+func isMatch[K any](target ottl.StringLikeGetter[K], pattern ottl.StringGetter[K]) (ottl.ExprFunc[K], error) {
+	literalPattern, ok := ottl.GetLiteralValue(pattern)
+	var compiledPattern *regexp.Regexp
+	var err error
+	if ok {
+		compiledPattern, err = regexp.Compile(literalPattern)
+		if err != nil {
+			return nil, fmt.Errorf(ottl.InvalidRegexErrMsg, "IsMatch", literalPattern, err)
+		}
 	}
 	return func(ctx context.Context, tCtx K) (any, error) {
+		cp := compiledPattern
+		if cp == nil {
+			patternVal, err := pattern.Get(ctx, tCtx)
+			if err != nil {
+				return nil, err
+			}
+			cp, err = regexp.Compile(patternVal)
+			if err != nil {
+				return nil, fmt.Errorf(ottl.InvalidRegexErrMsg, "IsMatch", patternVal, err)
+			}
+		}
```

Comment:
Is there a good way to abstract any of this? I get that Go is naturally verbose, but it's regrettable that we have to add so much code to each function to achieve this.

Reply:
adapted to your idea